site/content/blog/cue-and-carvel.md

```diff
@@ -0,0 +1,423 @@
+                name:  "simple-app"
+                image: "docker.io/dkalinin/k8s-simple-app@sha256:4c8b96d4fffdfae29258d94a22ae4ad1fe36139d47288b8960d9958d1e63a9d0"
+                env: [{
+                    name:  "HELLO_MSG"
+                    value: "stranger"
+                }]
+            }]
+        }
+    }
+}
+
+service: {
+    apiVersion: "v1"
+    kind:       "Service"
+    metadata: {
+        namespace: "default"
+        name:      "simple-app"
+    }
+    spec: {
+        ports: [{
+            port:       80
+            targetPort: 80
+        }]
+        selector: "simple-app": ""
+    }
+}
+```
+
+There are several things to note:
+- curly braces are explicitly used to contain maps (key-value structures)
+- brackets are explicitly used to specify collections
+- strings are always quoted which is unlike YAML where it's optional until you find ambiguity with other types like integers (this is one of my favorite choices in a configuration/data language -- i've even tried to experiment with [ytt's strict mode](https://carvel.dev/ytt/docs/v0.43.0/strict/) long ago to have a similar impact within YAML)
+- there is no notion of "documents" like in YAML -- root element of a file (and a package) is a map, hence if you want to include multiple Kubernetes resources, you'll have to find a "non-document" way to make them coexist (we'll dig into this next)
+- there is a single-line shorthand syntax to specify nested maps (see Deployment's `spec.selector` line)
+
+By the way, as you go along these examples feel free to use interactive [Cue playground](https://cuelang.org/play/#cue@export@cue) or [install Cue binary](https://github.com/cue-lang/cue#download-and-install) to run them locally. Also you may find it useful to refer to [Cue documentation](https://cuelang.org/docs/) as you read along for more in-depth explanation on specifics.
+
+To evaluate above example on the command line, run:
+
+```bash
+$ cue eval app.cue
+
+deployment: {
+    apiVersion: "apps/v1"
+    kind:       "Deployment"
+    metadata: {
+        namespace: "default"
+// ...snip...
+```
+
+Now that we can build basic configuration, let's see how to use it with something like kubectl or [Carvel's kapp](/kapp). Neither kubectl nor kapp will look inside top level keys like `service` and `deployment` which contain Kubernetes resources, yet somehow we need to let tools know what to deploy. Fortunately we can use somewhat "rare" Kubernetes resource `List` to combine multiple resources into one:
+
+`resources.cue`:
+```
+package app
+
+all: {
+    apiVersion: "v1"
+    kind: "List"
+    items: [deployment, service]
+}
+```
+
+Notice how contents of `deployment` and `service` fields were pulled in into `items` array. Cue allows to reference other parts of the data structure being built by field names (as long as it's unambigious) even if it's split across multiple files.
+
+```bash
+$ cue eval resources.cue app.cue --expression all --out yaml
```

Comment:
Just curious... any reason not to just switch to `cue export ...` from here on out? It would leave no ambiguity that what you've showed here is what `kapp-controller` does.

Reply:
it seems that cue export by default prints as json. i think it's bit nicer to use cue eval to print as cue by default.